File: src/com/medium/DesignCircularQueue_622.java
/**
 * 
 */
package com.medium;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class DesignCircularQueue_622 {
    // 2019年8月29日
    // Use array
    // Time Complexity: O(1)
    // Space Complexity:O(n)
    // Runtime: 44 ms, faster than 100.00%
    // Memory Usage: 38 MB, less than 28.57%

    private int rear;
    private int front;
    private int size;
    private int capacity;
    private int[] mQueue;

    /**
     * Initialize your data structure here. Set the size of the queue to be k.
     */
    public DesignCircularQueue_622(int k) {
        this.capacity = k;
        this.size = 0;
        this.rear = -1;
        this.front = -1;
        mQueue = new int[capacity];
    }

    /**
     * Insert an element into the circular queue. Return true if the operation
     * is successful.
     */
    public boolean enQueue(int value) {
        if (isFull()) {
            return false;
        }
        if(front == -1) {
            front++;
        }
        rear = (rear + 1) % capacity;
        mQueue[rear] = value;
        size++;
        return true;
    }

    /**
     * Delete an element from the circular queue. Return true if the operation
     * is successful.
     */
    public boolean deQueue() {
        if (isEmpty()) {
            return false;
        }
        front = (front + 1) % capacity;
        size--;
        return true;
    }

    /** Get the front item from the queue. */
    public int Front() {
        if (!isEmpty()) {
            return mQueue[front];
        }
        return -1;
    }

    /** Get the last item from the queue. */
    public int Rear() {
        if (!isEmpty()) {
            return mQueue[rear];
        }
        return -1;
    }

    /** Checks whether the circular queue is empty or not. */
    public boolean isEmpty() {
        return size == 0;
    }

    /** Checks whether the circular queue is full or not. */
    public boolean isFull() {
        return size == capacity;
    }
}

File: Golang_src/medium/designCircularQueue_622.go
package medium

// Use array:
// Time Complexity: O(1)
// Space Complexity:O(n)
// Runtime: 28 ms, faster than 7.61%
// Memory Usage: 7.2 MB, less than 31.52%
type MyCircularQueue struct {
	rear, front int
	size        int
	arr         []int
}

func ConstructorMyCircularQueue(k int) MyCircularQueue {
	return MyCircularQueue{
		size: k + 1,
		arr:  make([]int, k+1),
	}
}

func (this *MyCircularQueue) EnQueue(value int) bool {
	if this.IsFull() {
		return false
	}
	this.arr[this.rear] = value
	this.rear = (this.rear + 1) % this.size

	return true
}

func (this *MyCircularQueue) DeQueue() bool {
	if this.IsEmpty() {
		return false
	}
	this.front = (this.front + 1) % this.size

	return true
}

func (this *MyCircularQueue) Front() int {
	if this.IsEmpty() {
		return -1
	}
	return this.arr[this.front]
}

func (this *MyCircularQueue) Rear() int {
	if this.IsEmpty() {
		return -1
	}
	return this.arr[(this.rear-1+this.size)%this.size]
}

func (this *MyCircularQueue) IsEmpty() bool {
	return this.rear == this.front
}

func (this *MyCircularQueue) IsFull() bool {
	return (this.rear+1)%this.size == this.front
}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * obj := Constructor(k);
 * param_1 := obj.EnQueue(value);
 * param_2 := obj.DeQueue();
 * param_3 := obj.Front();
 * param_4 := obj.Rear();
 * param_5 := obj.IsEmpty();
 * param_6 := obj.IsFull();
 */
